qiskit/pulse/timeslots.py

```diff
@@ -150,7 +150,7 @@ def __repr__(self):
 class TimeslotCollection:
     """Collection of `Timeslot`s."""
 
-    def __init__(self, *timeslots: List[Timeslot]):
+    def __init__(self, *timeslots: Timeslot):
```

Comment:
As @lcapelluto pointed out, the inefficient seems to be instantiating `TimeslotCollection`. Could this be updated to use `is_mergeable_with`?

Reply:
If we could do this, then we shouldn't have to add `__merge`, which I'd like

Reply:
Do you mean we should change the interface as follows?
```
    def __init__(self, *timeslots: Tuple[Timeslot, ...]):
```

The inefficiency was in the full reconstruction of `_table` (and `_timeslots`) from scratch every time constructing a new TimeslotCollection. For example, when we add 1 time slot to existing (n-1) time slots, we constructed `_timeslots` from scratch that takes O(n) time, but this PR improve this to O(1) by using `__merged`.

Reply:
Public-facing merge methods should enforce a valid schedule. `merged` (should be `union`) should, therefore, ensure that the schedules to be merged do not overlap. Merging two timeslots should not be a two-step process for users (calling ismergeable and merge). 

If merging two `TimeslotCollections` merging two Schedules that were appended by default should be O(n_channels) performance if a sorted internal representation is used. As one only has to ensure that each channel starts after the last timeslot on the channel to which is being appended. 

I don't mind making TimeslotCollection's mutable as these are internal objects. However, the changes right now are confusing.

Reply:
I think it's reasonable that `Schedule` enforces a valid schedule, but I don't think that means `TimeslotCollections` should not allow overlap. I'm assuming users of `TimeslotCollection` are developers of schedule modules, who should know well when they do (or do not) have to check overlap. How about to rename to `_timeslots.py` to declare that it's not public for general users?
However, since current interface enforces non-overlap time slots, should I put the discussion on the spec of `TimeslotCollections` off later and keep current interface in this PR?

Reply:
In what case is an overlapping Timeslot valid? If Timeslots overlap, it may not be possible to make any nice guarantees about the underlying structure such as sorting of the Timeslot channel lists.